File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/modules/Ui/UI.java
package org.firstinspires.ftc.teamcode.modules.Ui;


import com.qualcomm.robotcore.hardware.Gamepad;

import org.firstinspires.ftc.robotcore.external.Telemetry;

/*
up - dpad up;
down - dpad down;
select/inc - dpad right;
dec - dpad left;
 */

public class UI {
    Telemetry telemetry;
    Gamepad gamepad;
    int itemCount = 0;
    int selectedItem = 1;

    boolean prevUp = false;
    boolean prevDown = false;
    boolean prevRight = false;
    boolean prevLeft = false;

    public UI(Telemetry telem, Gamepad _gamepad){
        gamepad = _gamepad;
        telemetry = telem;
    }

    public void update(){
        if(gamepad.dpad_up && !prevUp){
            selectedItem--;
            prevUp = true;
        }else if(!gamepad.dpad_up){
            prevUp = false;
        }
        if(gamepad.dpad_down && !prevDown){
            selectedItem++;
            prevDown = true;
        }else if(!gamepad.dpad_down){
            prevDown = false;
        }
        itemCount = 0;
    }

    public void resetSelected(){
        selectedItem = 1;
    }

    public boolean button(String label){
        itemCount++;
        addLine(label);
        if(selectedItem == itemCount){
            if(gamepad.dpad_right && !prevRight) {
                prevRight = true;
                return true;
            }else if(!gamepad.dpad_right){
                prevRight = false;
            }
        }
        return false;
    }

    public boolean floatInput(String label, FloatEx value, float step){
        itemCount++;
        addLine(label + ": " + value.value);
        if(selectedItem == itemCount) {
            if (gamepad.dpad_right && !prevRight) {
                prevRight = true;
                value.value += step;
                return true;
            } else if (!gamepad.dpad_right) {
                prevRight = false;
            }
            if (gamepad.dpad_left && !prevLeft) {
                prevLeft = true;
                value.value -= step;
                return true;
            } else if (!gamepad.dpad_left) {
                prevLeft = false;
            }
        }
        return false;
    }

    public boolean intInput(String label, IntEx value, int step){
        itemCount++;
        addLine(label + ": " + value.value);
        if(selectedItem == itemCount) {
            if (gamepad.dpad_right && !prevRight) {
                prevRight = true;
                value.value += step;
                return true;
            } else if (!gamepad.dpad_right) {
                prevRight = false;
            }
            if (gamepad.dpad_left && !prevLeft) {
                prevLeft = true;
                value.value -= step;
                return true;
            } else if (!gamepad.dpad_left) {
                prevLeft = false;
            }
        }
        return false;
    }

    public boolean checkbox(String label, BoolEx value){
        itemCount++;
        addLine(label + ": " + (value.value ? "[V]" : "[X]"));
        if(selectedItem == itemCount) {
            if (gamepad.dpad_right && !prevRight) {
                prevRight = true;
                value.value = !value.value;
                return true;
            } else if (!gamepad.dpad_right) {
                prevRight = false;
            }
        }
        return false;
    }

    public void label(String label){
        telemetry.addLine(label);
    }

    public void label(String label, Object arg){
        telemetry.addLine(label + ": " + arg.toString());
    }

    void addLine(String label){
        if(selectedItem == itemCount){
            telemetry.addLine("--" + label + "--");
        }else{
            telemetry.addLine(label);
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/modules/Ui/Ui.kt
package org.firstinspires.ftc.teamcode.modules.Ui

import com.qualcomm.robotcore.hardware.Gamepad
import org.firstinspires.ftc.robotcore.external.Telemetry

/*
up - dpad up;
down - dpad down;
select/inc - dpad right;
dec - dpad left;
*/
class UI
{
	private var telemetry: Telemetry? = null
	private var gamepad: Gamepad? = null
	private var itemCount = 0
	private var selectedItem = 1
	private var prevUp = false
	private var prevDown = false
	private var prevRight = false
	private var prevLeft = false
	
	constructor(telem: Telemetry, _gamepad: Gamepad)
	{
		init(telem, _gamepad)
	}
	
	constructor()
	{
	}
	
	fun init(telem: Telemetry, _gamepad: Gamepad)
	{
		gamepad = _gamepad
		telemetry = telem
	}
	
	fun update()
	{
		if(gamepad!!.dpad_up && !prevUp)
		{
			selectedItem--
			prevUp = true
		}
		else if(!gamepad!!.dpad_up)
		{
			prevUp = false
		}
		if(gamepad!!.dpad_down && !prevDown)
		{
			selectedItem++
			prevDown = true
		}
		else if(!gamepad!!.dpad_down)
		{
			prevDown = false
		}
		itemCount = 0
		telemetry?.update()
	}
	
	fun resetSelected()
	{
		selectedItem = 1
	}
	
	fun button(label: String): Boolean
	{
		itemCount++
		addLine(label)
		if(selectedItem == itemCount)
		{
			if(right() && !prevRight)
			{
				prevRight = true
				return true
			}
			else if(!right())
			{
				prevRight = false
			}
		}
		return false
	}
	
	fun floatInput(label: String, value: FloatPtr, step: Float): Boolean
	{
		itemCount++
		addLine(label + ": " + value.value)
		if(selectedItem == itemCount)
		{
			if(right() && !prevRight)
			{
				prevRight = true
				value.value += step
				return true
			}
			else if(!right())
			{
				prevRight = false
			}
			if(left() && !prevLeft)
			{
				prevLeft = true
				value.value -= step
				return true
			}
			else if(!left())
			{
				prevLeft = false
			}
		}
		return false
	}
	
	fun intInput(label: String, value: IntPtr, step: Int): Boolean
	{
		itemCount++
		addLine(label + ": " + value.value)
		if(selectedItem == itemCount)
		{
			if(right() && !prevRight)
			{
				prevRight = true
				value.value += step
				return true
			}
			else if(!right())
			{
				prevRight = false
			}
			if(left() && !prevLeft)
			{
				prevLeft = true
				value.value -= step
				return true
			}
			else if(!left())
			{
				prevLeft = false
			}
		}
		return false
	}
	
	fun checkbox(label: String, value: BoolPtr): Boolean
	{
		itemCount++
		addLine(label + ": " + if(value.value) "[V]" else "[X]")
		if(selectedItem == itemCount)
		{
			if(right() && !prevRight)
			{
				prevRight = true
				value.value = !value.value
				return true
			}
			else if(!right())
			{
				prevRight = false
			}
		}
		return false
	}
	
	fun label(label: String)
	{
		telemetry?.addLine("   $label")
	}
	
	fun label(label: String, arg: String?)
	{
		telemetry?.addLine("   $label: $arg")
	}
	
	fun addLine(label: String)
	{
		if(selectedItem == itemCount)
		{
			telemetry?.addLine(">>$label")
		}
		else
		{
			telemetry?.addLine("--$label")
		}
	}
	
	fun right(): Boolean
	{
		return gamepad!!.cross
	}
	
	fun left(): Boolean
	{
		return gamepad!!.triangle
	}
}